### PR TITLE
Refactored code to use AgentApplication and OmegaConf

### PR DESCRIPTION
**Pull request by patrick-halina**
I refactored the code with the goals of making the applications more configuration driven, and easier to compose out of modular components.

The refactor includes the following:
 * Replace RLApplication class with AgentApplication
 * Added ApplicationConfiguration using [OmegaConf](https://omegaconf.readthedocs.io/en/2.0_branch/) and dataclasses to make applications more configuration driven. Note that the hierarchy of configurations could change.
 * Cleaned up (most) docstring comments to standard reST format
 * Streamlined dqn, ddqn and ddpg examples to use generic configuration, AgentApplication and SimulationRunner

I didn't really design an abstraction for environments, for now the examples still use the TFRLApplication, with it's build_timesteps() function.

### CODE REVIEW

File: rl_bakery/agent_trainer/agents.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+from tf_agents.agents import ddpg
+from tf_agents.trajectories import time_step as ts
+import logging
+
+from rl_bakery.agent_trainer.config import Optimizer
+
+logger = logging.getLogger(__name__)
+
+
+class AgentTrainer(metaclass=abc.ABCMeta):
```
Comment (mehdi-ben-ayed) [inline rl_bakery/agent_trainer/agents.py:15]: Extracting the Agent initialization from RL_Application and placing it on its own is a good idea. 

Here are some suggestions in relation to improving this code:
* Rename the AgentTrainer to AgentInitializerInterface. This interface needs to be implemented to initial the different agents (e.g. dqn, sac...)

* Remove the _init_ function and make let the implementation of this interface handle that. Different types of agents might need different parameters

File: rl_bakery/agent_trainer/agents.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    def __init__(self, data_spec, config):
+        self._data_spec = data_spec
+        self._config = config
+
+    @abc.abstractmethod
+    def init_agent(self):
+        pass
+
+
+class TFAgentTrainer(AgentTrainer):
```
Comment (mehdi-ben-ayed) [inline rl_bakery/agent_trainer/agents.py:29]: I don't think inheritance is the right way to go here. Using it led to the creation of a long hierarchy of classes 
Level 1: TFAgentTrainer: Builds optimizer
Level 2: QAgent: Builds network 
Level 3: DQNAgent: Builds a DQN

Disadvantages of inheritance here:
* Coupling between the classes. If any code changes in level 1 or 2, all classes at level 3 will be impacted. 
* The behaviour inherited is set at compile-time. -> it would not be possible to change the parent class at runtime.  (e.g. the optimizer setup cannot change at runtime)

I think we should just have specific "AgentInitializers" be direct implementations of "AgentInitializersInterface". We could create two of these implementations: TFAgentDQNInitializer, and TFAgentDDPGInitializer for example.

File: rl_bakery/agent_trainer/agents.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+    @abc.abstractmethod
+    def _init_tfagent(self, optimizer):
+        pass
+
+    def _get_optimizer(self, optimizer_config):
+
+        # TODO use tf2 optimizers, support other optimizers
+        if optimizer_config.optimizer == Optimizer.Adam:
+            return tf.compat.v1.train.AdamOptimizer(learning_rate=optimizer_config.learning_rate)
```
Comment (mehdi-ben-ayed) [inline rl_bakery/agent_trainer/agents.py:42]: These optimizers have man other parameters that can be set which are not exposed in this setup such as _beta1_, _beta2_.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/adam.py#L41

File: rl_bakery/agent_trainer/agents.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+class QAgent(TFAgentTrainer):
+
+    def __init__(self, data_spec, config):
+        super().__init__(data_spec, config)
+
+    def _build_q_network(self):
+        # By default, this builds a q-network using the supplied fc_layer_params
+        # This function can be overridden to provide a more complicated Q Network
+        assert(self._config.agent.fc_layer_params)
+        q_net = q_network.QNetwork(self._data_spec.observation_spec, self._data_spec.action_spec,
```
Comment (mehdi-ben-ayed) [inline rl_bakery/agent_trainer/agents.py:63]: Same here in relation to the params of QNetwork such as dropout, etc.

https://github.com/tensorflow/agents/blob/master/tf_agents/networks/q_network.py#L46

File: rl_bakery/operation/train_agent_operation.py [hunk 5]
```diff
@@ -76,19 +85,18 @@ def _run(self, run_id, rl_app, engine_config, agent, timestep_df, **previous_tim
...
 
         # train the agent
         global_step = tf.compat.v1.train.get_or_create_global_step()
-        summary_interval = rl_app.training_config.get("summary_interval",
-                                                      rl_app.training_config.get("log_interval", 200))
+        summary_interval = self._application.config.project.summary_interval
         with tf.compat.v2.summary.record_if(lambda: tf.math.equal(global_step % summary_interval, 0)):
-            num_iterations = rl_app.training_config["num_iterations"]
-            mini_batch_size = rl_app.training_config["mini_batch_size"]
+            num_iterations = self._application.config.training.num_iterations
```
Comment (mehdi-ben-ayed) [inline rl_bakery/operation/train_agent_operation.py:98]: Accessing private members of the parent class here breaks the encapsulation. That is why the app is passed as a parameter in the method when called by the parent class.

File: rl_bakery/data_manager/builder.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+from rl_bakery.data_manager.data_manager import DataManager, DATANAME, InMemoryStorage, TFAgentStorage
+
+
+def build_inmemory_data_manager(application):
```
Comment (mehdi-ben-ayed) [inline rl_bakery/data_manager/builder.py:4]: Should we create a builder/factory to take care of the creation of the data_manager. That way it can support other data storage such as s3, etc...

File: rl_bakery/engine/timing_data.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+import logging
+
+logger = logging.getLogger(__name__)
+
+@dataclass
+class TimingData:
```
Comment (mehdi-ben-ayed) [inline rl_bakery/engine/timing_data.py:8]: I don't know if this class is needed. The configs are set in the RL App and the function to compute the current run_id is a utility.

File: rl_bakery/agent_trainer/agents.py (hunk 4 above)
Comment (patrick-halina) [reply rl_bakery/agent_trainer/agents.py]: it's okay, for more complex options they can override the function themselves

File: rl_bakery/operation/train_agent_operation.py (hunk 5 above)
Comment (patrick-halina) [reply rl_bakery/operation/train_agent_operation.py:98]: private is marked with two underscores, while a single underscore is 'protected', meaning derived classes have access to the values

File: rl_bakery/data_manager/builder.py (hunk 6 above)
Comment (patrick-halina) [reply rl_bakery/data_manager/builder.py:4]: yeah I agree, we should add a filesystem builder (which would work for hdfs, s3, gcp etc.)

**Review by mehdi-ben-ayed — APPROVED** [review]
